### PR TITLE
feat(stdlib): EdDSA sig verification

### LINKED ISSUE #1109: Add EdDSA verification to standard library
### Problem

Signature verification is an important building block for many protocols. The most popular in this domain being EcDSA. One problem with EcDSA is that it is not deterministic and so it is not suitable for usecases where one would want to use the signature as a nullifier.

### Proposed solution

EdDSA is a suitable alternative for this, moreover it is the alternative is the circomlib that many folks who are using circom will default to.

A link to EdDSA is here: https://github.com/iden3/circomlib/blob/master/circuits/eddsaposeidon.circom

Since we already have poseidon and Twisted edwards curve logic in our stdlib, we can implement the above by composing those two algorithms like in the circom library.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by shuklaayush**
# Related issue(s)

Resolves #1109
Supercedes #1136
Depends on #1312, #1343 

# Description

## Summary of changes

- Add a `eddsa::eddsa_poseidon_verify` function to stdlib 
- Update `std::ec` functions to use builtin `to_bits`
- Create `std::compat` with a `is_bn254` function

## Test additions / changes

# Checklist

- [x] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

**Comment by guipublic** [comment]
I implemented the to_bits for any field in PR #1343, could you compare the cost between your noir implementation vs the PR?
In my tests, it was only 2% better.

**Comment by shuklaayush** [comment]
I'm getting a `Error: could not satisfy all constraints` in this line with your changes

https://github.com/noir-lang/noir/blob/a04e78af4af8c37482d28621038496f4f5239c7b/crates/nargo_cli/tests/test_data/eddsa/src/main.nr#L57

**Comment by guipublic** [comment]
> I'm getting a `Error: could not satisfy all constraints` in this line with your changes

I don't think it is related to the to_bits, I checked the to_bits used and they produce the same version with the noir implementation and with my changes.

**Comment by shuklaayush** [comment]
> UPDATE: Using the ex you provided, I was able to see the issue and it is fixed

Thanks, everything seems to work now and total constraints are back to ~145k compared to ~247k constraints in https://github.com/noir-lang/noir/pull/1312#discussion_r1190930690

**Comment by shuklaayush** [comment]
@kevaundray This should be good for review

File: noir_stdlib/src/ec/swcurve.nr [hunk 1]
```diff
@@ -344,11 +344,20 @@ mod curvegroup {
 
         // Scalar multiplication (p + ... + p n times)
         fn mul(self, n: Field, p: Point) -> Point {
-            let n_as_bits = crate::ec::to_bits(n); // N_BITS-bit representation
+            let N_BITS = crate::field::modulus_num_bits();
 
-            self.bit_mul(n_as_bits, p)
```
Comment (guipublic) [inline noir_stdlib/src/ec/swcurve.nr:349]: Why don't you use a function for bit_mul?

File: noir_stdlib/src/field.nr [hunk 2]
```diff
@@ -53,3 +53,24 @@ fn modulus_be_bytes() -> [u8] {}
 
 #[builtin(modulus_le_bytes)]
 fn modulus_le_bytes() -> [u8] {}
+
+// Returns true if x is less than y
+// TODO: consider sftey of final few bits
+fn lt_bytes32(x: Field, y: Field) -> bool {
```
Comment (guipublic) [inline noir_stdlib/src/field.nr:59]: We do not allow comparison of field elements so this function must be removed from this file.

**Review by guipublic — DISMISSED** [review]

File: noir_stdlib/src/ec/swcurve.nr (hunk 1 above)
Comment (shuklaayush) [reply noir_stdlib/src/ec/swcurve.nr:349]: Yes, that would be ideal, but I get a panic if I change to this and do `nargo compile`:
```rust
// Scalar multiplication with scalar represented by a bit array (little-endian convention).
// If k is the natural number represented by `bits`, then this computes p + ... + p k times.
fn bit_mul<N>(self, bits: [u1; N], p: Point) -> Point {
    let mut out = Point::zero();
    let n = bits.len();
    
    for i in 0..n {
        out = self.add(
            self.add(out, out),
            if(bits[n - i - 1] == 0) {Point::zero()} else {p});
    }

    out
}

// Scalar multiplication (p + ... + p n times)
fn mul(self, n: Field, p: Point) -> Point {
    let N_BITS = crate::field::modulus_num_bits();
    let n_as_bits = n.to_le_bits(N_BITS as u32);

    self.bit_mul(n_as_bits, p)
}
```

File: noir_stdlib/src/field.nr (hunk 2 above)
Comment (shuklaayush) [reply noir_stdlib/src/field.nr:59]: One of the required conditions for checking the validity of an EdDSA signature is that the $s$ parameter should be less than the subgroup order of the curve over which the signature is defined (see [RFC8032 5.1.7](https://www.rfc-editor.org/rfc/rfc8032#page-14)). 
How do I check that if field element comparisons aren't allowed?

Comment (shuklaayush) [reply noir_stdlib/src/field.nr]: I moved it to `eddsa.nr`
Btw, what's the rationale behind not allowing field element comparisons in `field.nr`?

Comment (guipublic) [reply noir_stdlib/src/field.nr]: Thank you :)
Don't ask me this, I was pushing for having it!

File: noir_stdlib/src/ec/swcurve.nr (hunk 1 above)
Comment (guipublic) [reply noir_stdlib/src/ec/swcurve.nr:349]: Could you create an issue with the error you get?

Until the issue is resolved, you can try this which seems to work (and you would reference the issue):
```
            //temporary workaround until issue XXX is solved
            let mut n_as_bits :[u1;254] = [0;254];
            let tmp = n.to_le_bits(N_BITS as u32);
            for i in 0..254 {
               n_as_bits[i]= tmp[i]; 
            }
           self.bit_mul(n_as_bits, p)
```

File: noir_stdlib/src/field.nr (hunk 2 above)
Comment (shuklaayush) [reply noir_stdlib/src/field.nr]: Explanation for posterity: https://the.dev/posts/cryptography/circuit_design/integers_in_circ/
tldr: Finite fields don't have any notion of ordering

File: noir_stdlib/src/ec/swcurve.nr (hunk 1 above)
Comment (shuklaayush) [reply noir_stdlib/src/ec/swcurve.nr:349]: Okay, I did this. Though it adds a field-dependent constant that I don't like
Issue: #1354

File: noir_stdlib/src/field.nr (hunk 2 above)
Comment (guipublic) [reply noir_stdlib/src/field.nr]: Finite fields do have a notion of ordering as the simple 'less than' defines an ordering on the set.
However, is a 'set ordering', not a 'field ordering'.

**Review by guipublic — COMMENTED** [review]
My points have been addressed so for me this is OK to go, but I let final approval to Kev as I did not review the cryptography part.

File: noir_stdlib/src/ec/swcurve.nr [hunk 3]
```diff
@@ -344,11 +344,18 @@ mod curvegroup {
 
         // Scalar multiplication (p + ... + p n times)
         fn mul(self, n: Field, p: Point) -> Point {
-            let n_as_bits = crate::ec::to_bits(n); // N_BITS-bit representation
+            let N_BITS = crate::field::modulus_num_bits();
+
+            // temporary workaround until issue 1354 is solved
```
Comment (kevaundray) [inline noir_stdlib/src/ec/swcurve.nr:349]:
```suggestion
            // TODO: temporary workaround until issue 1354 is solved
```

File: crates/nargo_cli/tests/test_data/eddsa/src/main.nr [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        // the subgroup order. Note that choice of hash is flexible for this step.
+        // let r_a = hash::pedersen([_priv_key_a, msg])[0] % bjj.suborder; // modulus computed manually
+        let r_a = 1414770703199880747815475415092878800081323795074043628810774576767372531818;
+        // let r_b = hash::pedersen([_priv_key_b, msg])[0] % bjj.suborder; // modulus computed manually
+        let r_b = 571799555715456644614141527517766533395606396271089506978608487688924659618;
+
+        let r8_a = bjj.curve.mul(r_a, bjj.base8);
+        let r8_b = bjj.curve.mul(r_b, bjj.base8);
+
+        // let h_a: [Field; 6] = hash::poseidon::bn254::hash_5([
```
Comment (kevaundray) [inline crates/nargo_cli/tests/test_data/eddsa/src/main.nr:25]: Do we need this commented out code?

File: crates/nargo_cli/tests/test_data/eddsa/src/main.nr [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        //     r8_a.x,
+        //     r8_a.y,
+        //     pub_key_a.x,
+        //     pub_key_a.y,
+        //     msg,
+        // ]);
+
+        // let h_b: [Field; 6] = hash::poseidon::bn254::hash_5([
```
Comment (kevaundray) [inline crates/nargo_cli/tests/test_data/eddsa/src/main.nr:33]: Same question here

File: crates/nargo_cli/tests/test_data/eddsa/src/main.nr [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        //     r8_b.x,
+        //     r8_b.y,
+        //     pub_key_b.x,
+        //     pub_key_b.y,
+        //     msg,
+        // ]);
+
+        // let s_a = (r_a + _priv_key_a * h_a) % bjj.suborder; // modulus computed manually
```
Comment (kevaundray) [inline crates/nargo_cli/tests/test_data/eddsa/src/main.nr:41]: And for this section

File: noir_stdlib/src/ec/tecurve.nr [hunk 7]
```diff
@@ -364,17 +364,24 @@ mod curvegroup {
...
+
             out
         }
         
         // Scalar multiplication (p + ... + p n times)
         fn mul(self, n: Field, p: Point) -> Point {
-            let n_as_bits = crate::ec::to_bits(n); // N_BITS-bit representation
+            let N_BITS = crate::field::modulus_num_bits();
+
+            // temporary workaround until issue 1354 is solved
```
Comment (kevaundray) [inline noir_stdlib/src/ec/tecurve.nr:375]:
```suggestion
            // TODO: temporary workaround until issue 1354 is solved
```

**Review by kevaundray — COMMENTED** [review]
Left one question re commented out code

File: crates/nargo_cli/tests/test_data/eddsa/src/main.nr (hunk 4 above)
Comment (shuklaayush) [reply crates/nargo_cli/tests/test_data/eddsa/src/main.nr:25]: We don't need it, but I kept it as a hint of how the eddsa signing process works (and how the magic numbers below were generated). I'm okay with removing it if you think that's cleaner

I guess ideally you'd have standard test vectors to verify your implementation but I don't think those exist for pedersen based eddsa on bjj

**Review by kevaundray — APPROVED** [review]
